### PR TITLE
[ccw_passthrough_read_write] Add test case for read/write operations

### PR DESCRIPTION
**Pull request by smitterl**
Move helpers into module for reuse in new test cases.
Add a basic test case to confirm the attached DASD can be used.

Signed-off-by: Sebastian Mitterle <smitterl@redhat.com>

### CODE REVIEW

**Comment by smitterl** [comment]
@fbq815 Can you help review?

**Comment by smitterl** [comment]
```bash
# avocado run --vt-type libvirt --vt-machine-type s390-virtio --vt-connect-uri qemu:///system libvirt_ccw_passthrough.read_write
JOB ID     : f9943f6f3334a7af735e493e4edbbb188cde8708
JOB LOG    : /root/avocado/job-results/job-2021-06-29T11.37-f9943f6/job.log
 (1/1) type_specific.io-github-autotest-libvirt.libvirt_ccw_passthrough.read_write.happy_path: PASS (39.79 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 40.37 s
```

File: provider/vfio/ccw.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    Mounts device inside guest, identified by the chpids,
+    and runs some write/read operation.
+
+    The device is expected to be the only passed through device.
+    Per default the device gets a new filesystem setup.
+
+    :param session: logged in guest session
+    :param chipds: string representing CHPIDs, e.g. 11122122
+    :param makefs: if False, the device is expected to have a valid
+                   filesystem already
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:30]: docstring for the return

File: provider/vfio/ccw.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    """
+    device_id, subchannel_id = get_first_device_identifiers(chpids, session)
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:32]: If subchannel_id is not used, please use '_' instead.

File: provider/vfio/ccw.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        umount(session)
+    return True
+
+
+def read_write(session):
+    """
+    Writes, flushes and reads test file
+
+    :param session: guest session
+    """
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:51]: docstring for :raises ?

File: provider/vfio/ccw.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    mkfs(part, "ext3", session=session)
+
+
+def format_dasd(path, session):
+    """
+    Formats dasd disk and creates filesystem on path
+
+    :param path: dasd disk path, e.g. /dev/dasda
+    :param session: guest session
+    """
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:101]: :raises:

File: provider/vfio/ccw.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    if err:
+        raise TestError("Couldn't format disk. %s" % out)
+
+
+def umount(session):
+    """
+    Unmount dasd partition inside guest
+
+    :param session: guest session
+    """
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:114]: :raises:

File: provider/vfio/ccw.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+    if err:
+        raise TestError("Couldn't umount partition. %s" % out)
+
+
+def mount(session):
+    """
+    Mount dasd partition inside guest
+
+    :param session: guest session
+    """
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:127]: :raises:

File: provider/vfio/ccw.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        raise TestError("Couldn't mount partition. %s" % out)
+
+
+def set_device_online(device_id, session=None):
+    """
+    Sets device online
+
+    :param device_id: cssid.ssid.devno, e.g. 0.0.560a
+    :param session: guest session, command is run on host if None
+    """
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:141]: :raises:

File: provider/vfio/ccw.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        raise TestError("Could not set device online. %s" % out)
+
+
+def get_first_device_identifiers(chpids, session):
+    """
+    Gets the usual device identifier cssid.ssid.devno
+
+    :param chpids: chipids where the disk is connected, e.g. "11122122"
+    :param session: guest session
+    :returns pair of strings: "cssid.ssid.devno" "cssid.ssid.schid"
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:155]: :raises:

File: provider/vfio/ccw.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+
+def device_is_listed(session, chpids):
+    """
+    Checks if the css device is listed by comparing the channel
+    path ids.
+
+    :param session: guest console session
+    :param chipds: chpids where the disk is connected, e.g. "11122122"
+    """
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:176]: :return:

File: provider/vfio/ccw.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+                            if x[paths.HEADER["CHPIDs"]] == chpids]
+    return len(devices_inside_guest) > 0
+
+
+def set_override(schid):
+    """
+    Sets the driver override for the device' subchannel.
+
+    :param schid: Subchannel path id for device.
+    """
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:190]: :raises:

File: provider/vfio/ccw.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+    if err:
+        raise TestError("Can't set driver override. %s" % out)
+
+
+def unset_override(schid):
+    """
+    Unsets the driver override for the device' subchannel.
+
+    :param schid: Subchannel path id for device.
+    """
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:203]: :raises:

File: provider/vfio/ccw.py [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+        raise TestError("Can't set driver override. %s" % out)
+
+
+def start_device(uuid, schid):
+    """
+    Starts the mdev with mdevctl
+
+    :param uuid: device uuid
+    :param schid: subchannel id for the device
+    """
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:217]: :raises:

File: provider/vfio/ccw.py [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+    utils_package.package_install(["mdevctl",
+                                   "driverctl"])
+
+
+def get_device_info():
+    """
+    Gets the device info for passthrough
+    """
```
Comment (dzhengfy) [inline provider/vfio/ccw.py:268]: :return:

File: provider/vfio/ccw.py (hunk 1 above)
Comment (smitterl) [reply provider/vfio/ccw.py:30]: Done :)

File: provider/vfio/ccw.py (hunk 2 above)
Comment (smitterl) [reply provider/vfio/ccw.py]: Done :)

File: provider/vfio/ccw.py (hunk 3 above)
Comment (smitterl) [reply provider/vfio/ccw.py:53]: Done :)

File: provider/vfio/ccw.py (hunk 4 above)
Comment (smitterl) [reply provider/vfio/ccw.py:104]: Done :)

File: provider/vfio/ccw.py (hunk 5 above)
Comment (smitterl) [reply provider/vfio/ccw.py:118]: Done :)

File: provider/vfio/ccw.py (hunk 6 above)
Comment (smitterl) [reply provider/vfio/ccw.py:132]: Done :)

File: provider/vfio/ccw.py (hunk 7 above)
Comment (smitterl) [reply provider/vfio/ccw.py:147]: Done :)

File: provider/vfio/ccw.py (hunk 8 above)
Comment (smitterl) [reply provider/vfio/ccw.py:161]: Done :)

File: provider/vfio/ccw.py (hunk 9 above)
Comment (smitterl) [reply provider/vfio/ccw.py:182]: Done :)

File: provider/vfio/ccw.py (hunk 10 above)
Comment (smitterl) [reply provider/vfio/ccw.py:197]: Done :)

File: provider/vfio/ccw.py (hunk 11 above)
Comment (smitterl) [reply provider/vfio/ccw.py:211]: Done :)

File: provider/vfio/ccw.py (hunk 12 above)
Comment (smitterl) [reply provider/vfio/ccw.py:226]: Done :)

File: provider/vfio/ccw.py (hunk 13 above)
Comment (smitterl) [reply provider/vfio/ccw.py:280]: Done :)

File: provider/vfio/ccw.py (hunk 3 above)
Comment (smitterl) [reply provider/vfio/ccw.py:53]: Done :)

File: provider/vfio/ccw.py (hunk 4 above)
Comment (smitterl) [reply provider/vfio/ccw.py:104]: Done :)

File: provider/vfio/ccw.py (hunk 5 above)
Comment (smitterl) [reply provider/vfio/ccw.py:118]: Done :)

File: provider/vfio/ccw.py (hunk 6 above)
Comment (smitterl) [reply provider/vfio/ccw.py:132]: Done :)

File: provider/vfio/ccw.py (hunk 7 above)
Comment (smitterl) [reply provider/vfio/ccw.py:147]: Done :)

File: provider/vfio/ccw.py (hunk 10 above)
Comment (smitterl) [reply provider/vfio/ccw.py:197]: Done :)

File: provider/vfio/ccw.py (hunk 11 above)
Comment (smitterl) [reply provider/vfio/ccw.py:211]: Done :)

File: provider/vfio/ccw.py (hunk 12 above)
Comment (smitterl) [reply provider/vfio/ccw.py:226]: Done :)

File: provider/vfio/ccw.py (hunk 13 above)
Comment (smitterl) [reply provider/vfio/ccw.py:280]: Done :)

**Review by dzhengfy — COMMENTED** [review]
LGTM

**Review by dzhengfy — APPROVED** [review]

**Review by fbq815 — APPROVED** [review]

**Review by chunfuwen — APPROVED** [review]
LGTM

**Pull request merged**